nvaccess/nvda
# Add handling for accSelection returning IDispatch or negative child ID

### Link to issue number:

Fixes #13276

### Summary of the issue:

In case a single child is selected,
`IAccessible::get_accSelection` can either return
the child ID of a selected simple element or
an `IDispatch` for a selected accessible object, s.
https://docs.microsoft.com/en-us/windows/win32/api/oleacc/nf-oleacc-iaccessible-get_accselection
and
https://docs.microsoft.com/en-us/windows/win32/winauto/how-servers-implement-child-ids .

Previously, only positive child IDs were considered
in NVDA, while LibreOffice was returning negative child IDs
for Calc cells, which does not fit the MSAA spec.

NVDA did not yet support the case where an
`IDispatch` is returned.

### Description of how this pull request fixes the issue:

This PR implements the missing handling for an
`IDispatch` returned as accSelection in NVDA.
LibreOffice was adapted to return such a one from
LibreOffice 7.4 on, corresponding LibreOffice commit:
https://git.libreoffice.org/core/commit/00c0ee8cf0fac0c933c5ae600e99a64b1c7d4397

In addition, negative child IDs are now also accepted.
As pointed out by James Teh in the discussion on the
initial version of the pull request:

> The use of negative child ids doesn't fit well into the IAccessible spec, but
> it has been done by IAccessible2 for a very long time and should be considered
> standard for all intents and purposes. A negative child id should be treated as
> a unique id, while a positive child id should be treated as a child index.

Supporting negative child IDs fixes the issue for LibreOffice 7.3.

### Testing strategy:

Test the steps from issue #13276 and also make sure that the IDispatch case
works in NVDA when only the single cell A1 is selected:

1) start NVDA
2) start LibreOffice Calc 7.3
3) press Shift+Right to select cells A1 and B1
4) press Shift+Left to leave only cell A1 selected
5) verify that cell A1 is properly announced
6) set a breakpoint in the method changed by this PR and repeat 
   the previous steps with LibreOffice 7.4 (current development version) 
   or newer and verify that the IDispatch case is actually handled properly
   in NVDA, which is not the case without this PR

(For LibreOffice 7.4, announcement for the case described in
issue #13276 already works fine from a user perspective without
this PR, but just the LibreOffice change in place: An `IDispatch` is
returned by LibreOffice, the previous exception described in #13276
will no longer be triggered, but a `COMError` is triggered when querying
for the `IEnumVARIANT` interface a few lines below. Since that `COMError`
is handled in `getSelectedItemsCount` (in `source\NVDAObjects\IAccessible\__init__.py`)
and it falls back to calling the `nSelected` methods on the `IAccessibleTable2 interface`
instead, the user-facing end result is still OK.)

### Known issues with pull request:

none

### Change log entries:
*Bug fixes*
`When retrieving the count of selected children via accSelection, the case where a negative child ID or an IDispatch is returned by IAccessible::get_accSelection is now handled properly. (#13276)`

### Code Review Checklist:



  - description is up to date
  - change log entries
  - Unit tests
  - System (end to end) tests
  - Manual testing
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
